torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+    high_rank_tensors = [
+        tensor.view(tensor.shape[0], -1)
+        for tensor in tensors
+        if tensor.ndimension() > 1
+    ]
+    total_Ps_size = 0
+    ps_memory = None  # TODO(wayi): Store it in a dict of PowerState for warm-up.
+    total_Qs_size = 0
+    qs_memory = None  # TODO(wayi): Store it in a dict of PowerState for warm-up.
+    for tensor in high_rank_tensors:
+        n, m = tensor.shape
+        matrix_approximation_rank = min(n, m, state.matrix_approximation_rank)
+        total_Ps_size += n * matrix_approximation_rank
+        total_Qs_size += m * matrix_approximation_rank
+    ps_memory = torch.empty(total_Ps_size, device=device, dtype=dtype)
+    qs_memory = torch.empty(total_Qs_size, device=device, dtype=dtype)
+
+    # Create Ps and Qs that point to the allocated memory.
+    ps = []
+    qs = []
+    p_idx = 0
+    q_idx = 0
+    for tensor in high_rank_tensors:
+        n, m = tensor.shape
+        matrix_approximation_rank = min(n, m, state.matrix_approximation_rank)
+        ps.append(
+            ps_memory[p_idx : p_idx + n * matrix_approximation_rank].view(
+                n, matrix_approximation_rank
+            )
+        )
+        qs.append(
+            qs_memory[q_idx : q_idx + m * matrix_approximation_rank].view(
+                m, matrix_approximation_rank
+            )
+        )
+        p_idx += n * matrix_approximation_rank
+        q_idx += m * matrix_approximation_rank
+
+    # Initialize and then orthogonalize Qs.
+    with torch.random.fork_rng(devices=[]):
+        # Fork this RNG to avoid changing the seed globally and affecting the random sampling anywhere else in the training.
+        # The seed makes sure that the initial random values are the same across all the DDP replicas.
+        # Such seed should differ at every step.
+        # Since it is very slow to fork RNG state across all the CUDA devices,
+        # only fork on CPU and then move the generated tensor to the CUDA device.
+        torch.manual_seed(state.rng.randint(1_000_000_000))
+        for q in qs:
+            q.data = torch.randn(
+                *q.shape,
+                device="cpu",
+                dtype=dtype,
+            ).to(device)
+            _orthogonalize(q)
+
+    # Compute Ps.
+    for tensor, q, p in zip(high_rank_tensors, qs, ps):
+        torch.matmul(tensor, q, out=p)
+
+    # This allreduce is only applied to rank-1 tensors,
+    # so it should have been kicked off before the above computation on the high-rank tensors to hide more communication costs.
+    # However, this somehow requires a separate future chain at this time.
+    allreduce_contiguous_rank1_tensors_fut = dist.all_reduce(
+        rank1_tensors_memory, group=group_to_use, async_op=True
+    ).get_future()
+
+    def unpack_rank1_tensors_and_allreduce_ps(fut):
+        rank1_tensors_memory = fut.value()[0].div_(world_size)
```

Comment:
Is there a reason this isn't `fut.wait()`? Other calls seem to use `fut.wait()`.

Reply:
Good question!

`wait()` is already called once in the precursor callback, in the return statement, and it should only be called once.

`value()` means just reading the value without blocking, and it's user's responsibility to ensure proper wait before retrieving the value.

There is a recent PR that changed the semantics of value and wait. See: https://github.com/pytorch/pytorch/pull/48505/files#r532577873

I think I should change more `wait()` into `value()` in the original PowerSGD implementation. Previously, these two functions are kind of equivalent on NCCL backend.